src/net/tcp/socket.rs

```diff
@@ -106,6 +106,33 @@ impl TcpSocket {
         sys::tcp::set_linger(self.sys, dur)
     }
 
+    /// Sets whether keepalive messages are enabled to be sent on this socket.
+    ///
+    /// On Unix, this option will set the `SO_KEEPALIVE` as well as the
+    /// `TCP_KEEPALIVE` or `TCP_KEEPIDLE` option (depending on your platform).
```

Comment:
I've never been a fan of setting both options. `SO_KEEPALIVE` and `TCP_KEEPALIVE` are different, although related, options and should be able to be set separately. But I'm also ok with keeping this as is.

Reply:
Hmm, we could separate these and have a `set_keepalive(bool)` and a `set_keepalive_timeoutl(Duration)`... However, what behavior would we expect from `set_keepalive_timeout` if `SO_KEEPALIVE` is unset? Should it enable `SO_KEEPALIVE` first, or just return whatever error the OS sets in that case? 

I'm not immediately aware of what behavior we would expect from setting `TCP_KEEPIDLE`/`TCP_KEEPALIVE` when `SO_KEEPALIVE` is unset, but my guess is that it's OS-specific...

Reply:
I suppose we might also want a separate function to set the interval between keepalive probes...

Reply:
If we are going w/ socket options at this level, we *probably* should just expose them as they are provided by the OS?

Reply:
@carllerche just to make sure we're on the same page, you mean having separate methods for
- enabling/disabling keepalive (`SOL_SOCKET` + `SO_KEEPALIVE`)
- setting the keepalive time (`IPPROTO_TCP` + `TCP_KEEPIDLE`/`SO_KEEPALIVE`/`TCP_KEEPALIVE` depending on OS)
- setting the keepalive interval (`IPPROTO_TCP` + `TCP_KEEPINTVL` on Linux, i have no idea what this is on BSD...)
- setting the number of keepalive retries (`IPPROTO_TCP` + `TCP_KEEPCNT` on linux, not sure about BSD, winsock [doesn't support setting this](https://docs.microsoft.com/en-us/windows/win32/winsock/sio-keepalive-vals))

Reply:
I am not familiar enough w/ the portability details of the various options to comment. At the mio layer, we want to try to avoid being too opinionated. It's ok to have platform specific fns, but any given fn *should* behave the same across platforms.

Reply:
Okay, there are now separate API methods for setting each configurable value that the OS exposes. Let me know what you think!